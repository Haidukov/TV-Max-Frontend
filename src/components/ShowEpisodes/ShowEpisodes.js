import React from 'react';
import { useSelector } from 'react-redux';
import { useParams } from 'react-router';
import { Link } from 'react-router-dom';
import moment from 'moment';
import {
  getShowEpisodes,
  getShowEpisodesError,
  getShowEpisodesLoading
} from '../../store/modules/selectors';
import PageSpinner from '../Shared/PageSpinner/PageSpinner';
import ErrorMessage from '../Shared/ErrorMessage';

import s from './ShowEpisodes.scss';

const ShowEpisodes = () => {
  const { showId } = useParams();

  const episodes = useSelector(getShowEpisodes(showId));
  const isLoading = useSelector(getShowEpisodesLoading(showId));
  const error = useSelector(getShowEpisodesError(showId));

  const renderEpisode = episode => (
    <tr key={episode.id}>
      <td>{episode.number}</td>
      <td>
        <Link to={`/episodes/${episode.id}`}>
          {episode.name}
        </Link>
      </td>
      <td>{moment(episode.airdate, 'YYYY-MM-DD').format('LL')}</td>
    </tr>
  );

  if (isLoading) {
    return <PageSpinner />;
  }

  if (error) {
    return <ErrorMessage>Failed to load episodes list</ErrorMessage>;
  }

  return (
    <div>
      <table className={s.table}>
        <thead>
          <tr>
            <th>Number</th>
            <th>Name</th>
            <th>Date</th>
          </tr>
        </thead>
        <tbody>
          {episodes.map(renderEpisode)}
        </tbody>
      </table>
    </div>
  );
};

export default ShowEpisodes;
