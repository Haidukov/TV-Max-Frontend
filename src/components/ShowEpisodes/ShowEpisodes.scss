@import '../../assets/styles/variables.scss';


.table {
    margin-top: 10px;
    border-collapse: collapse;
    width: 100%;
    td, th {
        border: 1px solid $table-cell-1;
        padding: 8px;
    }

    tr {
        &:nth-child(even) {
            background-color: $table-cell-2;
        }
    
        &:hover {
            background-color: $table-cell-1;
        }
    }

    th {
        padding-top: 12px;
        padding-bottom: 12px;
        text-align: left;
        background-color: $table-header;
        color: white;
    }
}