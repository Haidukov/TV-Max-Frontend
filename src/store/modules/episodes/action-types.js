export const FETCH_EPISODES_REQUEST = 'FETCH_EPISODES_REQUEST';
export const FETCH_EPISODES_SUCCESS = 'FETCH_EPISODES_SUCCESS';
export const FETCH_EPISODES_ERROR = 'FETCH_EPISODES_ERROR';

export const FETCH_EPISODE_REQUEST = 'FETCH_EPISODE_REQUEST';
export const FETCH_EPISODE_SUCCESS = 'FETCH_EPISODE_SUCCESS';
export const FETCH_EPISODE_ERROR = 'FETCH_EPISODE_ERROR';
