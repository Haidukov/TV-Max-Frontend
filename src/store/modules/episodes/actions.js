import {
  FETCH_EPISODES_ERROR,
  FETCH_EPISODES_REQUEST,
  FETCH_EPISODES_SUCCESS,
  FETCH_EPISODE_ERROR,
  FETCH_EPISODE_REQUEST,
  FETCH_EPISODE_SUCCESS
} from '../action-types';

export const fetchEpisodesRequest = id => ({
  type: FETCH_EPISODES_REQUEST,
  payload: { id }
});

export const fetchEpisodesSuccess = (id, episodes) => ({
  type: FETCH_EPISODES_SUCCESS,
  payload: {
    id,
    episodes
  }
});

export const fetchEpisodesError = (id, error) => ({
  type: FETCH_EPISODES_ERROR,
  payload: {
    id,
    error
  }
});

export const fetchEpisodeRequest = id => ({
  type: FETCH_EPISODE_REQUEST,
  payload: { id }
});

export const fetchEpisodeSuccess = (id, data) => ({
  type: FETCH_EPISODE_SUCCESS,
  payload: {
    id,
    data
  }
});

export const fetchEpisodeError = (id, error) => ({
  type: FETCH_EPISODE_ERROR,
  payload: {
    id,
    error
  }
});
