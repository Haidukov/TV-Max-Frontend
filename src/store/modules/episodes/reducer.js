import produce from 'immer';
import {
  FETCH_EPISODES_ERROR,
  FETCH_EPISODES_REQUEST,
  FETCH_EPISODES_SUCCESS,
  FETCH_EPISODE_ERROR,
  FETCH_EPISODE_REQUEST,
  FETCH_EPISODE_SUCCESS
} from '../action-types';

const initialState = {
  byId: {},
  byShow: {}
};

const reducer = produce((draft, action) => {
  switch (action.type) {
    case FETCH_EPISODES_REQUEST: {
      if (!draft.byShow[action.payload.id]) {
        draft.byShow[action.payload.id] = {
          ids: []
        };
      }
      draft.byShow[action.payload.id].isLoading = true;
      draft.byShow[action.payload.id].error = null;
      break;
    }
    case FETCH_EPISODES_SUCCESS: {
      if (!draft.byShow[action.payload.id]) {
        draft.byShow[action.payload.id] = {
          ids: []
        };
      }
      draft.byShow[action.payload.id].isLoading = false;
      draft.byShow[action.payload.id].ids = [];
      action.payload.episodes.forEach(episode => {
        if (!draft.byId[episode.id]) {
          draft.byId[episode.id] = {};
        }
        draft.byId[episode.id].data = episode;
        draft.byShow[action.payload.id].ids.push(episode.id);
      });
      break;
    }
    case FETCH_EPISODES_ERROR: {
      if (!draft.byShow[action.payload.id]) {
        draft.byShow[action.payload.id] = {
          ids: []
        };
      }
      draft.byShow[action.payload.id].error = action.payload.error;
      draft.byShow[action.payload.id].isLoading = false;
      break;
    }

    case FETCH_EPISODE_REQUEST: {
      if (!draft.byId[action.payload.id]) {
        draft.byId[action.payload.id] = {};
      }
      draft.byId[action.payload.id].isLoading = true;
      draft.byId[action.payload.id].error = null;
      break;
    }

    case FETCH_EPISODE_SUCCESS: {
      if (!draft.byId[action.payload.id]) {
        draft.byId[action.payload.id] = {};
      }
      draft.byId[action.payload.id].isLoading = false;
      draft.byId[action.payload.id].data = action.payload.data;
      break;
    }

    case FETCH_EPISODE_ERROR: {
      if (!draft.byId[action.payload.id]) {
        draft.byId[action.payload.id] = {};
      }
      draft.byId[action.payload.id].isLoading = false;
      draft.byId[action.payload.id].error = action.payload.error;
      break;
    }
  }
}, initialState);

export default reducer;
