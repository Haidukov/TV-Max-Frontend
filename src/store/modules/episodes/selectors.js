import { createSelector } from 'reselect';
import { wrapSelector } from '../../../utils/selectors';

const getShowEpisodeIds = id => state => state.episodes.byShow[id]?.ids || [];

const getEpisodeByIdMap = state => state.episodes.byId;

export const getShowEpisodes = wrapSelector(id => createSelector(
  getShowEpisodeIds(id),
  getEpisodeByIdMap,
  (ids, byId) => ids?.map(episodeId => byId[episodeId]?.data)
));

export const getShowEpisodesLoading = id => state => !!state.episodes.byShow[id]?.isLoading;

export const getShowEpisodesError = id => state => state.episodes.byShow[id]?.error;

export const getEpisode = id => state => state.episodes.byId[id]?.data || {};

export const getEpisodeLoading = id => state => !!state.episodes.byId[id]?.isLoading;

export const getEpisodeError = id => state => state.episodes.byId[id]?.error;
