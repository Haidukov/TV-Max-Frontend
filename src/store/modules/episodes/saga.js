import {
  all,
  call,
  getContext,
  put,
  takeEvery
} from 'redux-saga/effects';
import { FETCH_EPISODES_REQUEST, FETCH_EPISODE_REQUEST } from './action-types';
import {
  fetchEpisodesError,
  fetchEpisodesSuccess,
  fetchEpisodeError,
  fetchEpisodeSuccess
} from '../actions';

function* fetchShowEpisodes({ payload }) {
  try {
    const api = yield getContext('api');
    const episodes = yield call(api.getShowEpisodes, payload.id);
    yield put(fetchEpisodesSuccess(payload.id, episodes));
  } catch (e) {
    yield put(fetchEpisodesError(payload.id, e));
  }
}

function* fetchEpisode({ payload }) {
  try {
    const api = yield getContext('api');
    const episode = yield call(api.getEpisode, payload.id);
    yield put(fetchEpisodeSuccess(payload.id, episode));
  } catch (e) {
    yield put(fetchEpisodeError(payload.id, e));
  }
}

export default function* saga() {
  yield all([
    takeEvery(
      FETCH_EPISODES_REQUEST,
      fetchShowEpisodes
    ),
    takeEvery(
      FETCH_EPISODE_REQUEST,
      fetchEpisode
    )
  ]);
}
