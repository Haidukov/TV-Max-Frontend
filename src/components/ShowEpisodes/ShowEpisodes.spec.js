import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter, Route } from 'react-router';
import { render } from '@testing-library/react';
import ShowEpisodes from './ShowEpisodes';
import configureStore from '../../store';
import episodes from '../../fixtures/episodes.json';

const LOADING_EPISODES_SHOW_ID = 1;
const FAILED_EPISODES_SHOW_ID = 2;
const LOADED_EPISODES_SHOW_ID = 3;

const renderComponent = id => {
  const preloadedState = {
    episodes: {
      byId: {
        [episodes[0].id]: {
          data: episodes[0]
        },
        [episodes[1].id]: {
          data: episodes[1]
        },
        [episodes[2].id]: {
          data: episodes[2]
        }
      },
      byShow: {
        [LOADING_EPISODES_SHOW_ID]: { // loading
          error: null,
          isLoading: true,
          ids: []
        },
        [FAILED_EPISODES_SHOW_ID]: { // error
          error: new Error('error'),
          isLoading: false,
          ids: []
        },
        [LOADED_EPISODES_SHOW_ID]: { // loaded
          error: null,
          isLoading: false,
          ids: [episodes[0].id, episodes[1].id, episodes[2].id]
        }
      }
    }
  };
  const store = configureStore(preloadedState);
  jest.spyOn(store, 'dispatch').mockImplementation(() => {});
  const wrapper = render(
    <MemoryRouter initialEntries={[`/path/${id}`]}>
      <Provider store={store}>
        <Route path='/path/:showId' component={ShowEpisodes} />
      </Provider>
    </MemoryRouter>
  );
  return wrapper;
};

const findSpinner = wrapper => wrapper.queryByRole('status');

const findErrorMessage = wrapper => wrapper.queryByRole('alert');

const findEpisodes = wrapper => wrapper.container.querySelectorAll('tbody tr');

const findLink = episodeEl => episodeEl.querySelector('a');

describe('<EpisodeepisodesPage />', () => {
  describe('When loading episode', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = renderComponent(LOADING_EPISODES_SHOW_ID);
    });

    it('should display loading indicator', () => {
      expect(findSpinner(wrapper)).toBeInTheDocument();
    });

    it('shouldn`t display error message', () => {
      expect(findErrorMessage(wrapper)).not.toBeInTheDocument();
    });

    it('shouldn`t display episodes', () => {
      expect(findEpisodes(wrapper).length).toBe(0);
    });

    it('should match snapshot', () => {
      expect(wrapper.asFragment()).toMatchSnapshot();
    });
  });

  describe('When error occured', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = renderComponent(FAILED_EPISODES_SHOW_ID);
    });

    it('shouldn`t display loading indicator', () => {
      expect(findSpinner(wrapper)).not.toBeInTheDocument();
    });

    it('should display error message', () => {
      expect(findErrorMessage(wrapper)).toBeInTheDocument();
    });

    it('shouldn`t display episodes', () => {
      expect(findEpisodes(wrapper).length).toBe(0);
    });

    it('should match snapshot', () => {
      expect(wrapper.asFragment()).toMatchSnapshot();
    });
  });

  describe('When data loaded', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = renderComponent(LOADED_EPISODES_SHOW_ID);
    });

    it('shouldn`t display loading indicator', () => {
      expect(findSpinner(wrapper)).not.toBeInTheDocument();
    });

    it('shouldn`t display error message', () => {
      expect(findErrorMessage(wrapper)).not.toBeInTheDocument();
    });

    it('should display episodes', () => {
      expect(findEpisodes(wrapper).length).toBe(3);
    });

    it('should link to episode details page', () => {
      const episodeEl = findEpisodes(wrapper)[0];
      const link = findLink(episodeEl);
      expect(link.getAttribute('href')).toBe(`/episodes/${episodes[0].id}`);
    });

    it('should match snapshot', () => {
      expect(wrapper.asFragment()).toMatchSnapshot();
    });
  });
});
