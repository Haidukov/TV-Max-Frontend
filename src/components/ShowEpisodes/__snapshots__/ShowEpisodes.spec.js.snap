// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EpisodeepisodesPage /> When data loaded should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <table
      class="table"
    >
      <thead>
        <tr>
          <th>
            Number
          </th>
          <th>
            Name
          </th>
          <th>
            Date
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            1
          </td>
          <td>
            <a
              href="/episodes/28"
            >
              Pilot
            </a>
          </td>
          <td>
            September 22, 2011
          </td>
        </tr>
        <tr>
          <td>
            2
          </td>
          <td>
            <a
              href="/episodes/29"
            >
              Ghosts
            </a>
          </td>
          <td>
            September 29, 2011
          </td>
        </tr>
        <tr>
          <td>
            3
          </td>
          <td>
            <a
              href="/episodes/30"
            >
              Mission Creep
            </a>
          </td>
          <td>
            October 6, 2011
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`<EpisodeepisodesPage /> When error occured should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="error"
    role="alert"
  >
    Failed to load episodes list
  </div>
</DocumentFragment>
`;

exports[`<EpisodeepisodesPage /> When loading episode should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="spinner"
    role="status"
  >
    <span
      class="css-1k4jg2i"
    />
  </div>
</DocumentFragment>
`;
